### PR TITLE
Add Joystick Support

### LINKED ISSUE #40: Implement gamepad/joystick support
use SDL/glfw's joystick input capability

### PR DESCRIPTION
**Pull request by rexim**
Resolves #40

### Features

- Assigning discrete actions to joystick buttons in `Settings > Controls`.
- Assigning joystick axis to mouse axis using `joystick_x` and `joystick_y` configuration variables in the console.

### Limitations

- Supported only a single joystick (the first one that was detected by SDL with index 0). Multi-joystick support could be implemented as a separate PR, maybe even by somebody else (not feeling it today).
- No UI for joystick/mouse axis mapping. You have to assign the [axis indices][SDL_JoystickGetAxis] to `joystick_x` and `joystick_y` configuration variables correspondingly. Could be also implemented separately by somebody who knows how to deal with settings UI code in [menus_callback.cpp] :)
- The [m_Joystick] object is never destroyed, because I could not find a good place for that. Please let me know if we should destroy it at all and where.

Also please let me know if you see any other problems with the PR.

[SDL_JoystickGetAxis]: https://wiki.libsdl.org/SDL_JoystickGetAxis
[menus_callback.cpp]: https://github.com/teeworlds/teeworlds/blob/4e14318cd1313f79040055fba9c3f57e9ca5b076/src/game/client/components/menus_callback.cpp
[m_Joystick]: https://github.com/tsoding/teeworlds/blob/bd4d5e047db6cd038598c61b77ab0e8f63832ca9/src/engine/client/input.cpp#L77

### CODE REVIEW

File: src/engine/client/input.h [hunk 1]
```diff
@@ -7,10 +7,13 @@ class CInput : public IEngineInput
 {
 	IEngineGraphics *m_pGraphics;
 	IConsole *m_pConsole;
+	SDL_Joystick *m_Joystick;
```
Comment (Dune-jr) [inline src/engine/client/input.h]: This should be `m_pJoystick` to fit Teeworlds codestyle (it's a pointer).

File: src/engine/shared/config_variables.h [hunk 2]
```diff
@@ -64,6 +64,9 @@ MACRO_CONFIG_INT(GfxUseX11XRandRWM, gfx_use_x11xrandr_wm, 1, 0, 1, CFGFLAG_SAVE|
 MACRO_CONFIG_INT(InpGrab, inp_grab, 0, 0, 1, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Disable OS mouse settings such as mouse acceleration, use raw mouse input mode")
 MACRO_CONFIG_INT(InpMousesens, inp_mousesens, 100, 5, 100000, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Ingame mouse sensitivity")
 
+MACRO_CONFIG_INT(JoystickAxisX, joystick_x, 0, 0, 6, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Joystick Axis that controls X axis of mouse")
+MACRO_CONFIG_INT(JoystickAxisY, joystick_y, 0, 0, 6, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Joystick Axis that controls Y axis of mouse")
```
Comment (Dune-jr) [inline src/engine/shared/config_variables.h]: Nitpick: preferably, the variable name should match the console variable name, so `JoystyickX` and `JoystickY`.

File: src/engine/client/input.cpp [hunk 3]
```diff
@@ -61,6 +63,33 @@ void CInput::Init()
...
+
+	if (!SDL_WasInit(SDL_INIT_JOYSTICK))
+	{
+		if (SDL_InitSubSystem(SDL_INIT_JOYSTICK) < 0) {
+			dbg_msg("joystick", "unable to init SDL joystick: %s", SDL_GetError());
+			return;
+		}
+	}
+
+	m_Joystick = NULL;
```
Comment (Dune-jr) [inline src/engine/client/input.cpp]: Teeworlds doesn't like `NULL`, I think you're supposed to use 0 (or 0x0). There are two other instances of that.

File: src/engine/client/input.cpp [hunk 4]
```diff
@@ -191,6 +228,69 @@ int CInput::Update()
 					Scancode = Event.key.keysym.scancode;
 					break;
 
+				// handle the joystick events
+				case SDL_JOYBUTTONUP:
+					Action = IInput::FLAG_RELEASE;
+
+					// fall through
+				case SDL_JOYBUTTONDOWN:
+					dbg_msg("joystick", "Joystick button down: %d", Event.jbutton.button);
```
Comment (Dune-jr) [inline src/engine/client/input.cpp]: This (and the `dbg_msg` below) looks quite verbose, right? I don't have a joystick to test but doing a print for every joystick event sounds too much.

File: src/engine/shared/config_variables.h [hunk 5]
```diff
@@ -64,6 +64,9 @@ MACRO_CONFIG_INT(GfxUseX11XRandRWM, gfx_use_x11xrandr_wm, 1, 0, 1, CFGFLAG_SAVE|
 MACRO_CONFIG_INT(InpGrab, inp_grab, 0, 0, 1, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Disable OS mouse settings such as mouse acceleration, use raw mouse input mode")
 MACRO_CONFIG_INT(InpMousesens, inp_mousesens, 100, 5, 100000, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Ingame mouse sensitivity")
 
+MACRO_CONFIG_INT(JoystickX, joystick_x, 0, 0, 6, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Joystick Axis that controls X axis of mouse")
+MACRO_CONFIG_INT(JoystickY, joystick_y, 0, 0, 6, CFGFLAG_SAVE|CFGFLAG_CLIENT, "Joystick Axis that controls Y axis of mouse")
```
Comment (oy) [inline src/engine/shared/config_variables.h]: The default value for the y axis should probably differ (1?) from the x one, so it works out of the box

**Comment by rexim** [comment]
@oy I closed the joystick in CInput's dtor and the application started to crash with double free on quit. I didn't research the cause, but I have a feeling that maybe CInput's dtor is called after SDL_Quit() or something. So, maybe it would be better to just not close the joystick and let SDL do its thing, I dunno.

**Comment by Dune-jr** [comment]
I attempted 
```cpp
	SDL_JoystickClose(m_pJoystick);
```
in `CInput::~CInput()` and it didn't crash, probably because I have no joystick; so I can't investigate.

**Comment by oy** [comment]
@rexim yeah, SDL probably cleans it up on shutdown. Then it's alright the way it is.

**Comment by EbrahimBGA** [comment]
issue

**Pull request merged**